File: go/internal/clock/vector_clock.go
package clock

// VectorClock maps peer IDs to counters
type VectorClock map[string]int64

// ComparisonResult is the relationship between two vector clocks
type ComparisonResult int

const (
	Equal ComparisonResult = iota
	Before
	After
	Concurrent
)

// Increment increments a peer counter on the vector clock
func Increment(clock VectorClock, peerID string) VectorClock {
	if clock == nil {
		clock = make(VectorClock)
	}
	clock[peerID] = clock[peerID] + 1
	return clock
}

// Merge two vector clocks (take max per peer)
func Merge(clock1, clock2 VectorClock) VectorClock {
	merged := make(VectorClock)
	for k, v := range clock1 {
		merged[k] = v
	}
	for k, v := range clock2 {
		if existing, ok := merged[k]; !ok || v > existing {
			merged[k] = v
		}
	}
	return merged
}

// Compare returns Equal|Before|After|Concurrent
func Compare(clock1, clock2 VectorClock) ComparisonResult {
	hasGreater, hasLess := false, false

	allKeys := make(map[string]struct{})
	for k := range clock1 {
		allKeys[k] = struct{}{}
	}
	for k := range clock2 {
		allKeys[k] = struct{}{}
	}

	for k := range allKeys {
		v1 := int64(0)
		v2 := int64(0)
		if vv, ok := clock1[k]; ok {
			v1 = vv
		}
		if vv, ok := clock2[k]; ok {
			v2 = vv
		}

		if v1 > v2 {
			hasGreater = true
		}
		if v1 < v2 {
			hasLess = true
		}
	}

	switch {
	case !hasGreater && !hasLess:
		return Equal
	case hasGreater && !hasLess:
		return After
	case hasLess && !hasGreater:
		return Before
	default:
		return Concurrent
	}
}

// HappensBefore returns true if clock1 is before or equal to clock2
func HappensBefore(clock1, clock2 VectorClock) bool {
	return Compare(clock1, clock2) == Before || Compare(clock1, clock2) == Equal
}

// NewVectorClock returns an empty clock
func NewVectorClock() VectorClock { return make(VectorClock) }

// Clone returns a shallow copy
func Clone(clock VectorClock) VectorClock {
	if clock == nil {
		return nil
	}
	copy := make(VectorClock, len(clock))
	for k, v := range clock {
		copy[k] = v
	}
	return copy
}

File: ts/src/components/clock/vector_clock.ts
// VectorClock maps peer IDs to counters
export type VectorClock = Record<string, number>;

// ComparisonResult is the relationship between two vector clocks
export enum ComparisonResult {
  Equal,
  Before,
  After,
  Concurrent,
}

// Increment increments a peer counter on the vector clock
export function increment(clock: VectorClock, peerID: string): VectorClock {
  if (!clock) {
    clock = {};
  }
  clock[peerID] = (clock[peerID] || 0) + 1;
  return clock;
}

// Merge two vector clocks (take max per peer)
export function merge(clock1: VectorClock, clock2: VectorClock): VectorClock {
  const merged: VectorClock = {};
  for (const k in clock1) {
    merged[k] = clock1[k];
  }
  for (const k in clock2) {
    if (!(k in merged) || clock2[k] > merged[k]) {
      merged[k] = clock2[k];
    }
  }
  return merged;
}

// Compare returns Equal|Before|After|Concurrent
export function compare(clock1: VectorClock, clock2: VectorClock): ComparisonResult {
  let hasGreater = false;
  let hasLess = false;

  const allKeys = new Set([...Object.keys(clock1), ...Object.keys(clock2)]);

  for (const k of allKeys) {
    const v1 = clock1[k] || 0;
    const v2 = clock2[k] || 0;

    if (v1 > v2) {
      hasGreater = true;
    }
    if (v1 < v2) {
      hasLess = true;
    }
  }

  if (!hasGreater && !hasLess) {
    return ComparisonResult.Equal;
  } else if (hasGreater && !hasLess) {
    return ComparisonResult.After;
  } else if (hasLess && !hasGreater) {
    return ComparisonResult.Before;
  } else {
    return ComparisonResult.Concurrent;
  }
}

// HappensBefore returns true if clock1 is before or equal to clock2
export function happensBefore(clock1: VectorClock, clock2: VectorClock): boolean {
  const cmp = compare(clock1, clock2);
  return cmp === ComparisonResult.Before || cmp === ComparisonResult.Equal;
}

// NewVectorClock returns an empty clock
export function newVectorClock(): VectorClock {
  return {};
}

// Clone returns a shallow copy
export function clone(clock: VectorClock): VectorClock {
  if (!clock) {
    return {};
  }
  return { ...clock };
}